modules/acton/src/restate/services/WebLeadVirtualObject.ts

```diff
@@ -0,0 +1,194 @@
+import type { UUID } from "node:crypto";
+import * as restate from "@restatedev/restate-sdk";
+import type { ExternalIntegrationState } from "common/external";
+import { serializeError } from "serialize-error";
+import { assert, is } from "tsafe";
+import { z } from "zod";
+import { fromError } from "zod-validation-error";
+import { SyncWithDB } from "../../db";
+import { WebLeadIntegrations } from "../../external/index";
+import {
+	type LeadState,
+	type WebFormLead,
+	WebFormLeadCreateRequestSchema,
+	type WebLead,
+} from "../../types";
+
+/**
+ * Helper function that runs before all of our exclusive handlers
+ * Handles initializing state from the database verifying assumptions
+ * @param ctx the restate Object context
+ * @param allowedStates a set of lead status states that we should allow this handler to run with
+ */
+async function setup(
+	ctx: restate.ObjectContext<LeadState>,
+	allowedStates: LeadState["Status"][],
+) {
+	const uuidParser = z.string().uuid();
+	if (!uuidParser.safeParse(ctx.key).success) {
+		throw new restate.TerminalError("Lead ID is not a valid UUIDv4", {
+			errorCode: 400,
+		});
+	}
+	if ((await ctx.get("Status")) === "ERROR") {
+		ctx.clearAll();
+	}
+	try {
+		await SyncWithDB(ctx, "RECEIVE");
+	} catch (e) {
+		await ctx.update((_) => ({
+			Status: "ERROR",
+			Error: serializeError(e),
+		}));
+		throw e;
+	}
+
+	const status = (await ctx.get("Status")) ?? "NONEXISTANT";
+	const canRun = allowedStates.includes(status);
+	if (!canRun) {
+		throw new restate.TerminalError(
+			`Lead ID '${ctx.key}' doesn't exist or is in incorrect status '${status}'`,
+			{ errorCode: 409 },
+		);
+	}
+}
+
+export const LeadVirtualObject = restate.object({
+	name: "WebLead",
+	handlers: {
+		status: restate.handlers.object.shared(
+			async (
+				ctx: restate.ObjectSharedContext<LeadState>,
+				req: Record<string, any>,
+			): Promise<LeadState> => {
+				//await CheckAuthorization(ctx);
```

Comment:
We'll need to set up auth for this service
```suggestion
				await CheckAuthorization(ctx);
```